File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.js
/**
 * @param {number[][]} points
 * @param {number} k
 * @return {number[][]}
 */

 /*
1. for each point,
    calculate the distance
    add to heap
2. pop k times from heap and store in res
3. return res
 */
var kClosest = function(points, k) {
    var heap = new MinPriorityQueue()
    var res = []

    for (var [x, y] of points) {
        var distance = Math.pow(x, 2) + Math.pow(y, 2)
        heap.enqueue([x,y], distance)
    }

    while (k > 0) {
        res.push(heap.dequeue().element)
        k--
    }
    return res
};

File: 0973-k-closest-points-to-origin/0973-k-closest-points-to-origin.py
class Solution:
    def kClosest(self, points: list[list[int]], k: int) -> list[list[int]]:
        maxHeap = []
        res = []

        heapq.heapify(maxHeap)

        for (x,y) in points:
            d = x ** 2 + y ** 2
            heapq.heappush(maxHeap, [-1 * d, [x, y]])

            if len(maxHeap) > k:
                heapq.heappop(maxHeap)

        for i in range(k):
            res.append(heapq.heappop(maxHeap)[1])

        return res
